### PR TITLE
Verify using sniffed packets instead of logs in link metric query and client capability query tests

### PR DESCRIPTION
**Pull request by aciliketcap**

### CODE REVIEW

**Comment by aciliketcap** [comment]
This PR is part of #1308, it utilizes sniffer in two more tests.

**Comment by aciliketcap** [comment]
Do we expect tests to fail at first issue (like most unit tests) or should they run until they can report each and every issue they encountered? Is it OK if they throw unhandled exceptions or should they mark themselves as failed and then log and swallow any expected exceptions?

I wrote the tests reporting as much issues as they can and never raising exceptions (that we expect).

**Comment by VladyslavTupikin** [comment]
Hi. @aciliketcap  Can you please prepare commits according to the instruction in CONTRIBUTING.md?

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py]: *Nitpick*: i really don't like such senseless variables.
```suggestion
        mid_cap_query = env.controller.dev_send_1905(env.agents[0].mac, 0x8009,
```
Will have more sense and easier to read, IMHO.

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py]: Same nitpick about sensless naming of variable.
```suggestion
        client_cap_query = self.check_cmdu_type_single("client capability query", 0x8009, env.controller.mac,
```

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py]: This block repeats frequently, should it be a separate method? It will significantly improve readability.

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py]: *Nitpick*: mainly, we tend to use `format` for string formatting.

**Review by abelog — CHANGES_REQUESTED** [review]
I still think that we should avoid repeat of code blocks. So not approving yet.
*Nitpicks* are optional to fix.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py]: I kept it as mid since the parameter name in check_cmdu_type method is mid and since it is not really used in test code.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py]: yeah, I guess I can make this more meaningful, like query_for_error instead of query1 and query_for_success instead of query2.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py]: All of them are removed since I'm using assert statements now.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py]: Oh, if most of the code is written that way I can change it too. But did you consider using format strings instead of .format() ? Many people (for example devs in my  current company :)) like format strings more.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: We can check a little more about these TLVs. E.g. at least the controller should be a neighbour, so it should be mentioned in both metrics.

File: tests/test_flows.py
Comment (arnout) [reply tests/test_flows.py]: I think "query" would be OK, but as soon as you need more than one, "query1" and "query2" are meaningless.

In this specific case, however, I think it would be better to split the test into two separate tests. The two parts of the test are pretty much independent of each other.

Comment (aciliketcap) [reply tests/test_flows.py]: I changed the variable names in that test. This time the lines are too long but I think it is OK for test code, what we are testing should be clear in variable names.

Btw I didn't used client_cap_query name since it is evident in the test name and check_cmdu_type error string parameter. I instead changed the name to reflect that we are testing for a query which returns a report with error and then a query with a successful report.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: Since you're changing this anyway, it would make sense to also move it down to the place where you actually need it (i.e. where the station associates).

**Comment by aciliketcap** [comment]
The CI flow gave flake8 error but it does not appear when I run flake8 with 'flake8 --config .flake8 tests/' ?

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: I don't like that you take `[0]` here - especially since the order of the TLVs is not defined. So either there is just one, and you should use check_cmdu_has_tlv_single, or there are several and you should iterate over them to find the one you want to check.

In this case, since we only requested a single station, there must be only a single client info tlv, so _single is appropriate.

Same issue occurs in a few other places.

File: tests/test_flows.py
Comment (arnout) [reply tests/test_flows.py]: Unfortunately, Intel is still using build machines with Python 3.4.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: I don't think we should print the full traceback. Also, I think it's worth repeating the test name.
```suggestion
                self.fail("Test {} has thrown an exception\n{}".format(test, e))
```

(Incidentally, this will also fix the flake8 error because `e` is now used)

With that, there's also no real need to handle the AssertionError differently.

BTW, it would be nice to show in the commit message some examples of what the output will be when an exception is handled. Something like:
```
For example, when a non-existing TLV field is accessed, the following failure will be printed:
...
```
For reviewers, that's much easier to help decide if it is done properly.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: I would actually move the assert inside the check_ function. Yes, there are some cases where we actually could continue the test even if the CMDU is missing, but that's not the common case. In the few cases where we do want to continue, we can still add a try block.

And actually, the assertion isn't even needed, because it's implied when you start using query_tlv below.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: This is not needed. We can pretty much assume that it's there, otherwise the tlv wouldn't exist to begin with. And if it doesn't, the following statement will simply raise an AttributeError.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: I don't think we should use asserts everywhere. It's better if we can continue the test, so assert isn't always good. And in this case I actually think the original code is easier to read.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: As noted elsewhere, this is not needed if we simply split the test into two tests.

Also, changes that are a result of review should typically be done as fixups. So in this case, do an interactive rebase and edit the commit that introduces mid1 and fix it there.

**Review by arnout — DISMISSED** [review]
Pretty good PR for your first contribution!

Myself, I would have put the commit that adds the exception handlers first, so the "use sniffed data" commits could immediately use the new approach. However, I kind of like this style as well, because it clearly shows why such a refactoring is needed - it's a bit of the [OneTwoThreeRefactor](http://c2.com/cgi/wiki?OneTwoThreeRefactor) approach.

Our review workflow is that I already approve now, even though there are still plenty of review comments to address. By approving already, you can simply make the requested changes and merge right away. And if there is something you're not sure of or disagree with, we discuss it either in the comments or in Slack before merging.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py]: Ah, I totally missed that this was causing the flake8 error.

Comment (aciliketcap) [reply tests/test_flows.py]: I expect tests to throw AssertionError when they fail and decide to quit. However I consider other exceptions as unexpected run-time exceptions and therefore I conclude that the test code has failed, not the DUT. I think we definitely need the stack trace (in debug / verbose mode) so that we can have a chance to figure out what went wrong without having to debug the test.

**Comment by abelog** [comment]
Removing requested changes mark as branch is rebuild by author.

**Review by abelog — DISMISSED** [review]

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py]: Then I'll definitely remove all format strings.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py]: I will be checking both links to controller and the other agent using what we know in env.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py]: I do that to prevent the AttributeError, test should only fail with AssertionError I think, otherwise we can't discern if the test failed or the test code failed.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: The True branch of the condition does nothing at all, so
```suggestion
                if not self.check_error:
                    self.fail("{}".format(ae))
```

However, I don't see why you would depend on check_error. If the assertion came after a completely unrelated failure, you still want to mark the failure. So I'd just do:

```suggestion
                if str(ae):
                    self.fail("{}".format(ae))
```

Comment (aciliketcap) [reply tests/test_flows.py]: I still wanted test to fail if self.fail() is not called before (inside test method) and there is no assertion failure message. I ended up printing a generic "Assertion failed" message in that case.

File: tests/test_flows.py
Comment (arnout) [reply tests/test_flows.py]: It doesn't actually prevent the attribute error - the assert will never even trigger if the attribute doesn't exist. You'd need
```
assert hasattr(query_tlv, "link_metric_query_type")
```

However, I think it is much more convenient if we can just access attributes without explicit asserts. In this case, we could actually implement a `__getattr__` function on TlvStruct that asserts.

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py]: *Nitpick*: please document arguments types. Makes usage of this method easier for others.

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py:126]: *Nitpick*: We tend to have commentaries above the string, not inline.

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py]: *Nitpick*: please document arguments types. Makes usage of this method easier for others.

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py]: To make message more informative.
```suggestion
                                      "Capability query was successful for disconnected STA")
```

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py]: Just small suggestion.
```suggestion
                    self.fail("Capability report does not contain expected frame")
```

File: tests/test_flows.py
Comment (abelog) [inline tests/test_flows.py:916]: Seems, those debug messages should be before `self.fail` call.

**Review by abelog — APPROVED** [review]
Nitpicks only, nothing crucial. In general - LGTM.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: (nitpick)
```suggestion
        assert result, "No CMDU {} found".format(msg)
```

You already had something like this below, so I guess it was just an oversight :-) But there are still other cases that could be simplified like this.

File: tests/test_flows.py
Comment (arnout) [reply tests/test_flows.py:126]: No, we have both actually.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py:189]: Hm, it would be nice if we had something like `ifdebug` that allows us to do:
```
assert len(tlvs) == 1, "More than one ({}) TLVs of type 0x{:02x} found{}".format(
    len(tlvs), tlv_type, ifdebug("\n  " + str(packet)))
```

Not for this PR, probably.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py]: If you change this in asserts as well, there's no need to correct the return type.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py:718]: Commit message is not correct... It says "use assert statements" but you're actually using the sniffer instead of check_log.

File: tests/test_flows.py
Comment (arnout) [inline tests/test_flows.py:913]: Doesn't safe_check_obj_attribute work here? Or is it just for the debug info that you have to special case it?

**Review by arnout — APPROVED** [review]
*Excellent* commit messages! And in general, this is a pretty good pull request.

In the future though I think it's better to merge things a little sooner and improve in follow-up pull requests. Review feedback doesn't necessarily mean that you have to make those changes right away.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py:916]: Yeah, but I though when you are reading you'd first want to be informed that the test has failed and then how it failed. If self.fail() made the test fail then of course I'd have to do it the other way around.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py:189]: We can prepare a string before hand (empty string if not debug) and then print it. Doing this in a separate function as you suggested is cleaner of course.

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py]: I did it. This method is not used btw :shrug:

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py:718]: You are absolutely right

File: tests/test_flows.py
Comment (aciliketcap) [reply tests/test_flows.py:913]: Yes, that's because I wanted to stop and print debug info if attribute exists but it is wrong. Maybe I can return False when safe_check_attribute fails and write it like;
if safe_check_obj_attrib():
  if cap_report.frame != expected_frame:
    debug_stuff()

**Pull request merged**
